src/solvers/dg/2d/dg.jl

```diff
@@ -1879,11 +1877,111 @@ end
 # Calculate and store fluxes across mortars (select correct method based on mortar type)
 calc_mortar_flux!(dg::Dg2D) = calc_mortar_flux!(dg, dg.mortar_type)
 
+# Calculate and store fluxes and nonconservative terms across L2 mortars
+calc_mortar_flux!(dg::Dg2D, mortar_type::Val{:l2}) = calc_mortar_flux!(dg.elements.surface_flux_values, dg, mortar_type,
+                                                                       have_nonconservative_terms(dg.equations), dg.l2mortars, dg.thread_cache)
+function calc_mortar_flux!(surface_flux_values, dg::Dg2D, mortar_type::Val{:l2}, nonconservative_terms::Val{true}, mortars, cache)
+  @unpack neighbor_ids, u_lower, u_upper, orientations = mortars
+  @unpack fstar_upper_threaded, fstar_lower_threaded = cache
+##
+#  FIXME! This is super hacked in for the MHD equations
+##
+  # Type alias only for convenience
+  A2d = MArray{Tuple{nvariables(dg), nnodes(dg)}, Float64}
+
+  noncons_diamond_primary_upper_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
+  noncons_diamond_primary_lower_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
+  noncons_diamond_secondary_upper_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
+  noncons_diamond_secondary_lower_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
+
+  Threads.@threads for m in 1:dg.n_l2mortars
+    large_element_id = dg.l2mortars.neighbor_ids[3, m]
+    upper_element_id = dg.l2mortars.neighbor_ids[2, m]
+    lower_element_id = dg.l2mortars.neighbor_ids[1, m]
+
+    # Choose thread-specific pre-allocated container
+    fstar_upper = fstar_upper_threaded[Threads.threadid()]
+    fstar_lower = fstar_lower_threaded[Threads.threadid()]
+
+    noncons_diamond_primary_upper = noncons_diamond_primary_upper_threaded[Threads.threadid()]
+    noncons_diamond_primary_lower = noncons_diamond_primary_lower_threaded[Threads.threadid()]
+    noncons_diamond_secondary_upper = noncons_diamond_secondary_upper_threaded[Threads.threadid()]
+    noncons_diamond_secondary_lower = noncons_diamond_secondary_lower_threaded[Threads.threadid()]
+
+    # Calculate fluxes
+    calc_fstar!(fstar_upper, u_upper, m, orientations, dg)
+    calc_fstar!(fstar_lower, u_lower, m, orientations, dg)
+
+    # Compute the nonconservative numerical terms along the upper and lower interface
+    # Done twice because left/right orientation matters
+    # 1 -> primary element and 2 -> secondary element
+    # See Bohm et al. 2018 for details on the nonconservative diamond "flux"
+    @views noncons_interface_flux!(noncons_diamond_primary_upper,
+                                   u_upper[1,:,:,:], u_upper[2,:,:,:],
+                                   m, nnodes(dg), orientations, equations(dg))
+    @views noncons_interface_flux!(noncons_diamond_secondary_upper,
+                                   u_upper[2,:,:,:], u_upper[1,:,:,:],
+                                   m, nnodes(dg), orientations, equations(dg))
```

Comment:
It seems like if `dg.l2mortars.large_sides[m] == 1` is true, you only ever need the `primary` fluxes, and otherwise only the `secondary` fluxes. However, you always calculate both primary and secondary fluxes here, no matter what. Maybe this can be an easy performance fix? In addition, that way you'd only need to allocate half of the thread caches.

Reply:
I struggled with this all of yesterday of what needs computed. The way I managed to treat it and get it working is that each `u_upper` and `u_lower` act like an interface from the conforming case. Because the non-conservative term is not unique at an interface I believe we always have to compute the four combinations between primary/secondary and upper/lower

Reply:
But I agree that the caching can probably be done in a cleaner way. Also I think we can rewrite `noncons_interface_flux` to be faster using the `get_surface_vals`functionality like in `calc_fstar!`. I am looking at this now

Reply:
>  Also I think we can rewrite `noncons_interface_flux` to be faster using the `get_surface_vals`functionality like in `calc_fstar!`.

:+1:

Reply:
> Because the non-conservative term is not unique at an interface I believe we always have to compute the four combinations between primary/secondary and upper/lower

Maybe I just don't get it, but if I grep for, e.g., `noncons_diamond_primary_upper`, after it is filled with the flux values I only ever see it being used where `larges_sides[m] == 1`:
![image](https://user-images.githubusercontent.com/3637659/92071778-f2fdcb00-edaf-11ea-8266-34b9dc6b4fa3.png)

Reply:
I can play around with moving the calls into if statements and the like. It will make the code repeat more and possibly harder to read (although I am just speculating here)

Reply:
I don't think it would be that bad - actually, a few lines can even be removed this way. I tried to do what I mean and the `parameters_mortar_alfven_wave.toml` test still works:
```diff
diff --git a/src/solvers/dg/2d/dg.jl b/src/solvers/dg/2d/dg.jl
index fec4798..05bde3c 100644
--- a/src/solvers/dg/2d/dg.jl
+++ b/src/solvers/dg/2d/dg.jl
@@ -1889,10 +1889,8 @@ function calc_mortar_flux!(surface_flux_values, dg::Dg2D, mortar_type::Val{:l2},
   # Type alias only for convenience
   A2d = MArray{Tuple{nvariables(dg), nnodes(dg)}, Float64}
 
-  noncons_diamond_primary_upper_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
-  noncons_diamond_primary_lower_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
-  noncons_diamond_secondary_upper_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
-  noncons_diamond_secondary_lower_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
+  noncons_diamond_upper_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
+  noncons_diamond_lower_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
 
   Threads.@threads for m in 1:dg.n_l2mortars
     large_element_id = dg.l2mortars.neighbor_ids[3, m]
@@ -1903,10 +1901,8 @@ function calc_mortar_flux!(surface_flux_values, dg::Dg2D, mortar_type::Val{:l2},
     fstar_upper = fstar_upper_threaded[Threads.threadid()]
     fstar_lower = fstar_lower_threaded[Threads.threadid()]
 
-    noncons_diamond_primary_upper = noncons_diamond_primary_upper_threaded[Threads.threadid()]
-    noncons_diamond_primary_lower = noncons_diamond_primary_lower_threaded[Threads.threadid()]
-    noncons_diamond_secondary_upper = noncons_diamond_secondary_upper_threaded[Threads.threadid()]
-    noncons_diamond_secondary_lower = noncons_diamond_secondary_lower_threaded[Threads.threadid()]
+    noncons_diamond_upper = noncons_diamond_upper_threaded[Threads.threadid()]
+    noncons_diamond_lower = noncons_diamond_lower_threaded[Threads.threadid()]
 
     # Calculate fluxes
     calc_fstar!(fstar_upper, u_upper, m, orientations, dg)
@@ -1916,46 +1912,50 @@ function calc_mortar_flux!(surface_flux_values, dg::Dg2D, mortar_type::Val{:l2},
     # Done twice because left/right orientation matters
     # 1 -> primary element and 2 -> secondary element
     # See Bohm et al. 2018 for details on the nonconservative diamond "flux"
-    @views noncons_interface_flux!(noncons_diamond_primary_upper,
-                                   u_upper[1,:,:,:], u_upper[2,:,:,:],
-                                   m, nnodes(dg), orientations, equations(dg))
-    @views noncons_interface_flux!(noncons_diamond_secondary_upper,
-                                   u_upper[2,:,:,:], u_upper[1,:,:,:],
-                                   m, nnodes(dg), orientations, equations(dg))
-    @views noncons_interface_flux!(noncons_diamond_primary_lower,
-                                   u_lower[1,:,:,:], u_lower[2,:,:,:],
-                                   m, nnodes(dg), orientations, equations(dg))
-    @views noncons_interface_flux!(noncons_diamond_secondary_lower,
-                                   u_lower[2,:,:,:], u_lower[1,:,:,:],
-                                   m, nnodes(dg), orientations, equations(dg))
+    if dg.l2mortars.large_sides[m] == 1 # -> small elements on right side
+      @views noncons_interface_flux!(noncons_diamond_upper,
+                                    u_upper[1,:,:,:], u_upper[2,:,:,:],
+                                    m, nnodes(dg), orientations, equations(dg))
+      @views noncons_interface_flux!(noncons_diamond_lower,
+                                    u_lower[1,:,:,:], u_lower[2,:,:,:],
+                                    m, nnodes(dg), orientations, equations(dg))
+    else # large_sides[m] == 2 -> small elements on left side
+      @views noncons_interface_flux!(noncons_diamond_upper,
+                                    u_upper[2,:,:,:], u_upper[1,:,:,:],
+                                    m, nnodes(dg), orientations, equations(dg))
+      @views noncons_interface_flux!(noncons_diamond_lower,
+                                    u_lower[2,:,:,:], u_lower[1,:,:,:],
+                                    m, nnodes(dg), orientations, equations(dg))
+    end
+
     # Copy flux small to small
     if dg.l2mortars.large_sides[m] == 1 # -> small elements on right side
       if dg.l2mortars.orientations[m] == 1
         # L2 mortars in x-direction
-        surface_flux_values[:, :, 1, upper_element_id] .= (fstar_upper + noncons_diamond_primary_upper)
-        surface_flux_values[:, :, 1, lower_element_id] .= (fstar_lower + noncons_diamond_primary_lower)
+        surface_flux_values[:, :, 1, upper_element_id] .= (fstar_upper + noncons_diamond_upper)
+        surface_flux_values[:, :, 1, lower_element_id] .= (fstar_lower + noncons_diamond_lower)
       else
         # L2 mortars in y-direction
-        surface_flux_values[:, :, 3, upper_element_id] .= (fstar_upper + noncons_diamond_primary_upper)
-        surface_flux_values[:, :, 3, lower_element_id] .= (fstar_lower + noncons_diamond_primary_lower)
+        surface_flux_values[:, :, 3, upper_element_id] .= (fstar_upper + noncons_diamond_upper)
+        surface_flux_values[:, :, 3, lower_element_id] .= (fstar_lower + noncons_diamond_lower)
       end
     else # large_sides[m] == 2 -> small elements on left side
       if dg.l2mortars.orientations[m] == 1
         # L2 mortars in x-direction
-        surface_flux_values[:, :, 2, upper_element_id] .= (fstar_upper + noncons_diamond_secondary_upper)
-        surface_flux_values[:, :, 2, lower_element_id] .= (fstar_lower + noncons_diamond_secondary_lower)
+        surface_flux_values[:, :, 2, upper_element_id] .= (fstar_upper + noncons_diamond_upper)
+        surface_flux_values[:, :, 2, lower_element_id] .= (fstar_lower + noncons_diamond_lower)
       else
         # L2 mortars in y-direction
-        surface_flux_values[:, :, 4, upper_element_id] .= (fstar_upper + noncons_diamond_secondary_upper)
-        surface_flux_values[:, :, 4, lower_element_id] .= (fstar_lower + noncons_diamond_secondary_lower)
+        surface_flux_values[:, :, 4, upper_element_id] .= (fstar_upper + noncons_diamond_upper)
+        surface_flux_values[:, :, 4, lower_element_id] .= (fstar_lower + noncons_diamond_lower)
       end
     end
 
     # Project small fluxes to large element
     for v in 1:nvariables(dg)
       if dg.l2mortars.large_sides[m] == 1 # -> large element on left side
-        @views large_surface_flux_values = (dg.l2mortar_reverse_upper * (fstar_upper[v, :] + noncons_diamond_primary_upper[v,:]) +
-                                            dg.l2mortar_reverse_lower * (fstar_lower[v, :] + noncons_diamond_primary_lower[v,:]))
+        @views large_surface_flux_values = (dg.l2mortar_reverse_upper * (fstar_upper[v, :] + noncons_diamond_upper[v,:]) +
+                                            dg.l2mortar_reverse_lower * (fstar_lower[v, :] + noncons_diamond_lower[v,:]))
         if dg.l2mortars.orientations[m] == 1
           # L2 mortars in x-direction
           surface_flux_values[v, :, 2, large_element_id] .= large_surface_flux_values
@@ -1964,8 +1964,8 @@ function calc_mortar_flux!(surface_flux_values, dg::Dg2D, mortar_type::Val{:l2},
           surface_flux_values[v, :, 4, large_element_id] .= large_surface_flux_values
         end
       else # large_sides[m] == 2 -> large element on right side
-        @views large_surface_flux_values = (dg.l2mortar_reverse_upper * (fstar_upper[v, :] + noncons_diamond_secondary_upper[v,:]) +
-                                            dg.l2mortar_reverse_lower * (fstar_lower[v, :] + noncons_diamond_secondary_lower[v,:]))
+        @views large_surface_flux_values = (dg.l2mortar_reverse_upper * (fstar_upper[v, :] + noncons_diamond_upper[v,:]) +
+                                            dg.l2mortar_reverse_lower * (fstar_lower[v, :] + noncons_diamond_lower[v,:]))
         if dg.l2mortars.orientations[m] == 1
           # L2 mortars in x-direction
           surface_flux_values[v, :, 1, large_element_id] .= large_surface_flux_values
```

Reply:
You are right. I put in the pointwise version of this and the EOCs are identical.